haskell/gen_ghc_bindist.py

```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env python
+
+# This is a happy-path tool to download the bindist
+# download paths and hashes, for maintainers.
+# It uses the hashes provided by download.haskell.org.
+
+from __future__ import print_function
+
+import pprint
+import sys
+import urllib2
+
+# all GHC versions we generate.
+VERSIONS = [
+    "8.6.4",
+    "8.6.3",
+    "8.6.2",
+    "8.4.4",
+    "8.4.3",
+    "8.4.2",
+    "8.4.1",
+    "8.2.2",
+    "8.0.2",
+    "7.10.3",
+]
+
+# All architectures we generate.
+# bazel: bazel name
+# upstream: download.haskell.org name
+ARCHES = [
+    { "bazel": "linux_amd64",
+      "upstream": "x86_64-deb8-linux", },
+    { "bazel": "darwin_amd64",
+      "upstream": "x86_64-apple-darwin" },
+    { "bazel": "windows_amd64",
+      "upstream": "x86_64-unknown-mingw32" },
+]
+
+
+# An url to a bindist tarball.
+def link_for_tarball(arch, version):
+    return "https://downloads.haskell.org/~ghc/{ver}/ghc-{ver}-{arch}.tar.xz".format(
+        ver = version,
+        arch = arch,
+    )
+
+# An url to a version’s tarball hashsum file.
+# The files contain the hashsums for all arches.
+def link_for_sha256_file(version):
+    return "https://downloads.haskell.org/~ghc/{ver}/SHA256SUMS".format(
+        ver = version
+    )
+
+# Parses the tarball hashsum file for a version.
+def parse_sha256_file(content, version):
+    res = {}
+    for line in content:
+        # f5763983a26dedd88b65a0b17267359a3981b83a642569b26334423f684f8b8c  ./ghc-8.4.3-i386-deb8-linux.tar.xz
+        hash_file = line.strip().split("  ./")
+        assert len(hash_file) == 2
+        hash = hash_file[0]
+        file_ = hash_file[1]
+        prefix = "ghc-{ver}-".format(ver = version)
+        suffix = ".tar.xz"
+        # ghc-8.4.3-i386-deb8-linux.tar.xz
+        if file_.startswith(prefix) and file_.endswith(suffix):
+            # i386-deb8-linux
+            name = file_[len(prefix):-len(suffix)]
+            res[name] = hash
+        else:
+            eprint("Info: Can't parse the sha256 field for {ver}: {entry}".format(ver = version, entry = line.strip()))
```

Comment:
Is this supposed to fail? `eprint` does not.

Reply:
No, just an info field, because some of the releases have special files. Though you are right, it might be more correct to fail in these cases.

Reply:
Okay, I changed it so that it fails for everything that was not explicitely filtered out.